src/app/server/JointFabricDatastorage.h

```diff
@@ -0,0 +1,198 @@
+    {
+        Set(nodeId, fabricId, state, label);
+    }
+
+    GenericDatastoreNodeInformationEntry(const GenericDatastoreNodeInformationEntry & op) { *this = op; }
+
+    GenericDatastoreNodeInformationEntry & operator=(const GenericDatastoreNodeInformationEntry & op)
+    {
+        Set(op.nodeID, op.fabricIndex, op.commissioningStatusEntry.state, MakeOptional(op.friendlyName));
+        return *this;
+    }
+
+    void Set(NodeId nodeId, FabricIndex fabricId, Clusters::JointFabricDatastore::DatastoreStateEnum state,
+             Optional<CharSpan> label = NullOptional)
+    {
+        SetFabricIndex(fabricId);
+        this->nodeID                         = nodeId;
+        this->commissioningStatusEntry.state = state;
+        Set(label);
+    }
+
+    void Set(Optional<CharSpan> label = NullOptional)
+    {
+        if (label.HasValue())
+        {
+            memset(mFriendlyNameBuffer, 0, sizeof(mFriendlyNameBuffer));
+            if (label.Value().size() > sizeof(mFriendlyNameBuffer))
+            {
+                memcpy(mFriendlyNameBuffer, label.Value().data(), sizeof(mFriendlyNameBuffer));
+                this->friendlyName = CharSpan(mFriendlyNameBuffer, sizeof(mFriendlyNameBuffer));
+            }
+            else
+            {
+                memcpy(mFriendlyNameBuffer, label.Value().data(), label.Value().size());
+                this->friendlyName = CharSpan(mFriendlyNameBuffer, label.Value().size());
+            }
+        }
+        else
+        {
+            this->friendlyName = CharSpan();
+        }
+    }
+
+private:
+    static constexpr size_t kFriendlyNameMaxSize = 32u;
+
+    char mFriendlyNameBuffer[kFriendlyNameMaxSize];
+};
+
+class JointFabricDatastorage
+{
+public:
+    static JointFabricDatastorage & GetInstance()
+    {
+        static JointFabricDatastorage sInstance;
+        return sInstance;
+    }
+
+    CHIP_ERROR SetAdministratorFabricIndex(FabricIndex fabricIndex)
+    {
+        mAdministratorFabricIndex = fabricIndex;
+        return CHIP_NO_ERROR;
+    }
+    FabricIndex GetAdministratorFabricIndex() { return mAdministratorFabricIndex; }
+
+    CHIP_ERROR SetAnchorNodeId(NodeId anchorNodeId)
+    {
+        mAnchorNodeId = anchorNodeId;
+        return CHIP_NO_ERROR;
+    }
+    NodeId GetAnchorNodeId() { return mAnchorNodeId; }
+
+    CHIP_ERROR SetAnchorVendorId(VendorId anchorVendorId)
+    {
+        mAnchorVendorId = anchorVendorId;
+        return CHIP_NO_ERROR;
+    }
+    VendorId GetAnchorVendorId() { return mAnchorVendorId; }
+
+    CHIP_ERROR AddPendingNode(FabricIndex fabricId, NodeId nodeId, const CharSpan & friendlyName);
+    CHIP_ERROR UpdateNode(NodeId nodeId, const CharSpan & friendlyName);
+    CHIP_ERROR RemoveNode(NodeId nodeId);
+    CHIP_ERROR RefreshNode(NodeId nodeId);
+
+    CHIP_ERROR SetNode(NodeId nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum state);
+
+    CHIP_ERROR AddGroupKeySetEntry(uint16_t groupKeySetId);
```

Comment:
do we need this method? I think it might not be needed.

Reply:
Addressed in the latest commit.